drivers/can/can_numaker.c

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright (c) 2023 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_numaker_canfd
+
+#include <zephyr/drivers/reset.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/can.h>
+#include <zephyr/drivers/can/can_mcan.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/clock_control_numaker.h>
+#include <zephyr/logging/log.h>
+#include <soc.h>
+#include <NuMicro.h>
+
+LOG_MODULE_REGISTER(can_numaker, CONFIG_CAN_LOG_LEVEL);
+
+/* Implementation notes
+ * 1. Use Bosch M_CAN driver (m_can) as backend
+ *    NOTE: m_can only supports single configuration of max element numbers for all instances
+ * 2. Need to modify can_numaker_get_core_clock() for new SOC support
+ */
+
+struct can_numaker_config {
+	mm_reg_t   canfd_base;
+	mem_addr_t  mrba;
+	mem_addr_t  mram;
+	const struct reset_dt_spec reset;
+	uint32_t    clk_modidx;
+	uint32_t    clk_src;
+	uint32_t    clk_div;
+	const struct device *clk_dev;
+	void (*irq_config_func)(const struct device *dev);
+	const struct pinctrl_dev_config *pincfg;
+};
+
+struct can_numaker_data {
+	/* Avoid zero length struct */
+	uint32_t resv0;
+};
+
+static int can_numaker_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	const struct can_mcan_config *mcan_config = dev->config;
+	const struct can_numaker_config *config = mcan_config->custom;
+	int rc = 0;
```

Comment:
no need to initialize this

Reply:
Last time, you guided me the better way.
Applied it: no need variable "rc", return directly as:
```
	if (clksrc_rate_idx >= ARRAY_SIZE(clksrc_rate_tab)) {
		LOG_ERR("Invalid clock source rate table or index");
		return -EIO;
	}
    ......
   return 0;

```